pkg/image/containerdregistry/resolver.go

```diff
@@ -71,71 +37,47 @@ func NewResolver(configDir string, skipTlSVerify, plainHTTP bool, roots *x509.Ce
 	return docker.NewResolver(opts), nil
 }
 
-func credential(cfg *configfile.ConfigFile) func(string) (string, string, error) {
-	return func(hostname string) (string, string, error) {
-		hostname = resolveHostname(hostname)
-		auth, err := cfg.GetAuthConfig(hostname)
-		if err != nil {
-			return "", "", err
-		}
-		if auth.IdentityToken != "" {
-			return "", auth.IdentityToken, nil
-		}
-		if auth.Username == "" && auth.Password == "" {
-			return "", "", nil
-		}
-
-		return auth.Username, auth.Password, nil
+func credentialFunc(configDir, repo string) func(string) (string, string, error) {
+	if configDir == "" {
+		configDir = dockerconfig.Dir()
 	}
-}
-
-// protects against a data race inside the docker CLI
-// TODO: upstream issue for 20.10.x is tracked here https://github.com/docker/cli/pull/3410
-// newer versions already contain the fix
-var configMutex sync.Mutex
 
-func loadConfig(dir string) (*configfile.ConfigFile, error) {
-	configMutex.Lock()
-	defer configMutex.Unlock()
-
-	if dir == "" {
-		dir = config.Dir()
+	// By default, we will use the docker config file in the standard docker config directory.
+	// However, if REGISTRY_AUTH_FILE or DOCKER_CONFIG environment variables are set, we will
+	// use those (in that order) instead to derive the auth config file.
+	authFile := filepath.Join(configDir, dockerconfig.ConfigFileName)
+	if defaultAuthFile := auth.GetDefaultAuthFile(); defaultAuthFile != "" {
+		authFile = defaultAuthFile
 	}
 
-	dockerConfigJSON := filepath.Join(dir, config.ConfigFileName)
-	cfg := configfile.New(dockerConfigJSON)
+	// We don't use the function parameter in the credential function we return because containerd
+	// only passes in the hostname. Instead, we will use our repo parameter to get the credentials
+	// using the repo-aware GetCredentials function.
+	return func(_ string) (string, string, error) {
+		var (
+			cred types.DockerAuthConfig
+			err  error
+		)
+
+		// In order to maintain backward-compatibility with the original credential getter,
+		// we will first try to get the credentials from the auth config file we derived above,
+		// if it exists.
+		if stat, statErr := os.Stat(authFile); statErr == nil && stat.Mode().IsRegular() {
+			cred, err = config.GetCredentials(&types.SystemContext{AuthFilePath: authFile}, repo)
```

Comment:
This err is unchecked and can get overwritten on line 72

Reply:
Either option to make it obvious it's intentional would be good :)

Reply:
Went with option 2.